### ISSUE TITLE
tap sometimes prints "unknown" messages

### ISSUE THREAD

**Issue by hawkw** [linkerd/linkerd2#1293]
After running `tap` against a test running in GCP for a period of time, `tap` began printing "unknown" events:
```
rsp id=0:9131 proxy=in  src=10.48.13.98:58050 dst=10.48.13.97:9090 tls=disabled :status=200 latency=534µs
end id=0:9131 proxy=in  src=10.48.13.98:58050 dst=10.48.13.97:9090 tls=disabled grpc-status=OK duration=33µs response-length=66B
rsp id=0:9131 proxy=in  src=10.48.13.98:38972 dst=10.48.12.176:9090 tls=disabled :status=200 latency=754µs
end id=0:9131 proxy=in  src=10.48.13.98:38972 dst=10.48.12.176:9090 tls=disabled grpc-status=Unknown duration=0µs response-length=0B
req id=0:9131 proxy=in  src=10.48.13.98:43350 dst=10.48.11.230:8080 tls=disabled :method=POST :authority=t3-svc:8080 :path=/
rsp id=0:9131 proxy=in  src=10.48.13.98:43350 dst=10.48.11.230:8080 tls=disabled :status=200 latency=742µs
end id=0:9131 proxy=in  src=10.48.13.98:43350 dst=10.48.11.230:8080 tls=disabled duration=30µs response-length=87B
rsp id=0:9131 proxy=out src=10.48.12.177:37098 dst=gateway-58b775f798-lzh78:8080 tls=no_identity :status=500 latency=5169µs
end id=0:9131 proxy=out src=10.48.12.177:37098 dst=gateway-58b775f798-lzh78:8080 tls=no_identity duration=25µs response-length=164B
unknown proxy=??? src=0.0.0.0:0 dst=0.0.0.0:0 tls=
unknown proxy=??? src=0.0.0.0:0 dst=0.0.0.0:0 tls=
unknown proxy=??? src=0.0.0.0:0 dst=0.0.0.0:0 tls=
unknown proxy=??? src=0.0.0.0:0 dst=0.0.0.0:0 tls=
unknown proxy=??? src=0.0.0.0:0 dst=0.0.0.0:0 tls=
unknown proxy=??? src=0.0.0.0:0 dst=0.0.0.0:0 tls=
unknown proxy=??? src=0.0.0.0:0 dst=0.0.0.0:0 tls=
unknown proxy=??? src=0.0.0.0:0 dst=0.0.0.0:0 tls=
unknown proxy=??? src=0.0.0.0:0 dst=0.0.0.0:0 tls=
```

Running the same `tap` command again produced normal output. The tap in question was running for over 30 minutes, but @rmars reports that she's seen similar output immediately after running `tap`.

The logs from the control plane's `tap` container seem normal. A scattering of 
```
level=error msg="rpc error: code = Canceled desc = context canceled"
```
messages are logged, but these seem to be logged during normal operation.

<details>
<summary>The `tap` command and version:</summary>

Command: 
```
 ➜ bin/conduit tap -c liza-conduit ns/liza-conduit-tap-test
```

CLI version:
```
 ➜ bin/conduit version
Client version: git-ba2e13c7
Server version: git-219b017c
```
</details>

<details>
<summary>tap server logs (truncated):</summary>

```
➜ stern -n liza-conduit controller -c tap -t --color always
+ controller-6df56785c9-vr4qw › tap
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:53.787228885Z time="2018-07-06T21:29:53Z" level=info msg="running conduit version v0.5.0"
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:53.788647954Z time="2018-07-06T21:29:53Z" level=info msg="starting admin server on :9998"
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:53.788659094Z time="2018-07-06T21:29:53Z" level=info msg="starting gRPC server on 127.0.0.1:8088"
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:53.789029351Z time="2018-07-06T21:29:53Z" level=info msg="waiting for caches to sync"
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:53.789628612Z E0706 21:29:53.789410       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Pod: Get https://10.51.240.1:443/api/v1/pods?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:53.789642989Z E0706 21:29:53.789490       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.ReplicationController: Get https://10.51.240.1:443/api/v1/replicationcontrollers?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:53.789647142Z E0706 21:29:53.789512       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Namespace: Get https://10.51.240.1:443/api/v1/namespaces?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:53.789650843Z E0706 21:29:53.789554       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1beta2.Deployment: Get https://10.51.240.1:443/apis/apps/v1beta2/deployments?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:53.790104423Z E0706 21:29:53.789905       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Service: Get https://10.51.240.1:443/api/v1/services?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:54.790593989Z E0706 21:29:54.790320       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Pod: Get https://10.51.240.1:443/api/v1/pods?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:54.791183793Z E0706 21:29:54.790916       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.ReplicationController: Get https://10.51.240.1:443/api/v1/replicationcontrollers?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:54.792274973Z E0706 21:29:54.791949       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Namespace: Get https://10.51.240.1:443/api/v1/namespaces?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:54.794008558Z E0706 21:29:54.793691       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1beta2.Deployment: Get https://10.51.240.1:443/apis/apps/v1beta2/deployments?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:54.79480186Z E0706 21:29:54.794528       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Service: Get https://10.51.240.1:443/api/v1/services?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:55.791386451Z E0706 21:29:55.791004       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Pod: Get https://10.51.240.1:443/api/v1/pods?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:55.792710195Z E0706 21:29:55.792210       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.ReplicationController: Get https://10.51.240.1:443/api/v1/replicationcontrollers?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:55.794343876Z E0706 21:29:55.793270       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Namespace: Get https://10.51.240.1:443/api/v1/namespaces?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:55.796185976Z E0706 21:29:55.794895       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1beta2.Deployment: Get https://10.51.240.1:443/apis/apps/v1beta2/deployments?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:55.796674064Z E0706 21:29:55.796406       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Service: Get https://10.51.240.1:443/api/v1/services?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:56.792151289Z E0706 21:29:56.791706       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Pod: Get https://10.51.240.1:443/api/v1/pods?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:56.792791944Z E0706 21:29:56.792550       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.ReplicationController: Get https://10.51.240.1:443/api/v1/replicationcontrollers?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:56.794491631Z E0706 21:29:56.794041       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Namespace: Get https://10.51.240.1:443/api/v1/namespaces?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:56.795633323Z E0706 21:29:56.795322       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1beta2.Deployment: Get https://10.51.240.1:443/apis/apps/v1beta2/deployments?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:56.797189631Z E0706 21:29:56.796828       1 reflector.go:205] k8s.io/client-go/informers/factory.go:87: Failed to list *v1.Service: Get https://10.51.240.1:443/api/v1/services?limit=500&resourceVersion=0: dial tcp 10.51.240.1:443: connect: connection refused
controller-6df56785c9-vr4qw tap 2018-07-06T21:29:57.889198644Z time="2018-07-06T21:29:57Z" level=info msg="caches synced"
controller-6df56785c9-vr4qw tap 2018-07-06T21:30:58.453869788Z time="2018-07-06T21:30:58Z" level=info msg="Tapping 1 pods for target: {Namespace:liza-conduit-tap-test Type:deployments Name:gateway}"
```
</details>

**Comment by grampelberg** [comment]
@hawkw this has happened to me in the past when the underlying TCP connection dies. `unknown` is basically a tight loop around trying to fetch a new message and having it come back with an error immediately.

The old behavior was to panic (because of a pointer reference later on).

**Comment by hawkw** [comment]
@grampelberg Thanks for shining light into what's going on here! In that case, I think this behaviour is certainly more correct than panicking, but perhaps the tap client ought to print a message indicating that the TCP connection failed so that it's clearer what's going on?

**Comment by grampelberg** [comment]
In a perfect world, we'd just go back to connecting and try that again. I think it is perfectly reasonable to quit the process and ask users to start it up again though.

**Comment by hawkw** [comment]
As long as it's clear why it happened, I think the process exiting would be fine too.

**Comment by grampelberg** [comment]
Related issue: #1287

**Comment by adleong** [comment]
Fixed by #1371
